File: Fortran/ex5/linearRegCostFunction.f90
	subroutine linearRegCostFunction(M, N, X, y, theta, lambda, Jcost, grad)
	implicit none
	
	integer::M, N
	double precision,dimension(M):: y, h
	double precision,dimension(M, N)::X
	double precision,dimension(N)::theta, grad
	double precision::lambda, Jcost 
	integer::i,j
	
	
	Jcost = 0
	do i =1, M
		h(i) = 0.
		do j = 1, N
			h(i) = h(i) + X(i,j)*theta(j)
		enddo
		
		Jcost = Jcost + (h(i)-y(i))**2./M/2.
	enddo
	
	do j=2,N
		Jcost = Jcost + theta(j)**2 *lambda/M/2.
	enddo
	
	
		grad = 0.
	do i=1, M
		grad(1) = grad(1)+ (h(i)-y(i))*X(i,1)/M
	enddo
		
	do j=2,N
		do i=1, M
		grad(j) = grad(j)+ (h(i)-y(i))*X(i,j)/M + theta(j)*lambda/M
		enddo
	enddo
	
	end subroutine linearRegCostFunction

File: C/ex5/linearRegCostFunction.c
#include <stdio.h>

double linearRegCostFunction(int M,int N,double X[M][N],double y[M],double theta[N], double lambda, double grad[N])
{
	double Jcost;
	int i, j;
	double h[M];
	
	Jcost = 0.f;
	for(i=0;i<M;i++){
			h[i] = 0.f;
		for(j=0;j<N;j++){
			h[i] +=  X[i][j]*theta[j];
		}
//		printf("h[%d]= %lf  %lf\n", i+1, h[i], y[i]);
		
		Jcost = Jcost + (h[i] - y[i])*(h[i] - y[i])/M/2.;	
		
//		printf("Jcost[%d]= %lf \n", i+1, Jcost);	
	}		

	for(j=1;j<N;j++){
		Jcost = Jcost + theta[j]*theta[j]*lambda/M/2.;
	}
	
	
	
	
	for(j=0;j<N;j++)
		grad[j] = 0;
					
	for(i=0;i<M;i++){
		grad[0] = grad[0] + (h[i]-y[i])*X[i][0]/M ;
	}
			
	for(j=1;j<N;j++){
		for(i=0;i<M;i++){
			grad[j] = grad[j] + (h[i]-y[i])*X[i][j]/M + lambda/M*theta[j];
		}
	}	
	
	
	return Jcost;
}
